src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,170 @@
+        Node<T> node = getNode(index);
+        return node.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        validateIndexForGet(index);
+        Node<T> node = getNode(index);
+        T oldValue = node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateIndexForGet(index);
+        Node<T> removedNode;
+        if (index == 0) {
+            removedNode = head;
+            head = head.next;
+            if (head == null) {
+                tail = null;
+            }
+        } else {
+            Node<T> previousNode = getNode(index - NUMBER_TO_SUBTRACT);
+            removedNode = previousNode.next;
+            previousNode.next = removedNode.next;
+            if (index == size - NUMBER_TO_SUBTRACT) {
+                tail = previousNode;
+            }
+        }
+        size--;
+        return removedNode.value;
     }
 
     @Override
-    public boolean remove(T object) {
+    public boolean remove(T value) {
+        Node<T> current = head;
+        Node<T> previous = null;
+        while (current != null) {
+            if ((value == null && current.value == null)
+                    || (value != null && value.equals(current.value))) {
+                unlink(previous, current);
+                size--;
+                return true;
+            }
+            previous = current;
+            current = current.next;
+        }
         return false;
     }
 
+    private void unlink(Node<T> previous, Node<T> current) {
+        if (previous == null) {
+            head = current.next;
+            if (head == null) {
+                tail = null;
+            }
+        } else {
+            previous.next = current.next;
+            if (current.next == null) {
+                tail = previous;
+            }
+        }
+    }
+
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void addFirst(Node<T> newNode) {
+        newNode.next = head;
+        head = newNode;
+        if (tail == null) {
+            tail = newNode;
+        }
+        size++;
+    }
+
+    private void addLast(Node<T> newNode) {
+        if (tail == null) {
+            head = newNode;
+        } else {
+            tail.next = newNode;
+        }
+        tail = newNode;
+        size++;
+    }
+
+    private Node<T> getNode(int index) {
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        return current;
+    }
+
+    private void validateIndexForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException(INDEX_FOR_EXCEPTION
+                    + index + SIZE_FOR_EXCEPTION + size);
+        }
+    }
```

Comment:
Probably, if you using this code one time, you can add it to the add((T value, int index) method and don't need to create additional